File: packages/lynx-maps/ios/SigxMapMarkerUI.swift
import Foundation
import UIKit
import MapKit
import Lynx

/// Native UI for the `<sigx-map-marker>` JSX element.
///
/// A marker doesn't render its own real view — Google Maps and MapKit both
/// model pins as data attached to the map. We still need a `LynxUI<UIView>`
/// because Lynx's UI tree expects every child to be a UI. The native view is
/// a zero-size hidden `UIView` that never gets laid out: the parent
/// `SigxMapUI` intercepts `insertChild` for markers and skips calling super,
/// so the dummy view is created but never attached to the map view hierarchy.
///
/// Prop surface (v1):
///   - `coordinate`  → JSON-stringified `{ latitude, longitude }`
///   - `title`       → callout title
///   - `description` → callout subtitle
///   - `marker-id`   → forwarded as `event.detail.id` on `bindmarkerpress`
// Class is NOT marked `@objc` — Swift forbids that on generic subclasses
// of an ObjC lightweight-generic type like `LynxUI<__covariant V>`. Member-
// level `@objc` / `@objc(name)` annotations still bridge because `LynxUI`
// itself is `@objc`, so `__lynx_prop_config__*` discovery still works.
public class SigxMapMarkerUI: LynxUI<UIView> {
    internal weak var owningMap: SigxMapUI?

    internal var currentCoordinate = CLLocationCoordinate2D(latitude: 0, longitude: 0)
    internal var currentTitle: String?
    internal var currentSubtitle: String?
    internal var markerId: String = ""

    public override func createView() -> UIView? {
        let v = UIView(frame: .zero)
        v.isHidden = true
        return v
    }

    // Explicit prop-setter registration. See SigxWebViewUI for rationale.
    @objc public class func propSetterLookUp() -> NSArray {
        return [
            ["coordinate", "setCoordinate:requestReset:"],
            ["title", "setTitle:requestReset:"],
            ["description", "setDescription:requestReset:"],
            ["marker-id", "setMarkerId:requestReset:"],
        ] as NSArray
    }

    internal func makeAnnotation() -> MKPointAnnotation {
        let a = MKPointAnnotation()
        a.coordinate = currentCoordinate
        a.title = currentTitle
        a.subtitle = currentSubtitle
        return a
    }

    // MARK: - Prop setters

    @objc public func setCoordinate(_ value: NSString?, requestReset: Bool) {
        guard let raw = value as String?, !raw.isEmpty,
              let data = raw.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let lat = (obj["latitude"] as? NSNumber)?.doubleValue,
              let lon = (obj["longitude"] as? NSNumber)?.doubleValue
        else { return }
        currentCoordinate = CLLocationCoordinate2D(latitude: lat, longitude: lon)
        owningMap?.markerDidUpdate(self)
    }

    @objc(__lynx_prop_config__coordinate)
    public class func __lynxPropConfigCoordinate() -> [String] {
        return ["coordinate", "setCoordinate", "NSString *"]
    }

    @objc public func setTitle(_ value: NSString?, requestReset: Bool) {
        currentTitle = value as String?
        owningMap?.markerDidUpdate(self)
    }

    @objc(__lynx_prop_config__title)
    public class func __lynxPropConfigTitle() -> [String] {
        return ["title", "setTitle", "NSString *"]
    }

    @objc public func setDescription(_ value: NSString?, requestReset: Bool) {
        currentSubtitle = value as String?
        owningMap?.markerDidUpdate(self)
    }

    @objc(__lynx_prop_config__description)
    public class func __lynxPropConfigDescription() -> [String] {
        return ["description", "setDescription", "NSString *"]
    }

    @objc public func setMarkerId(_ value: NSString?, requestReset: Bool) {
        markerId = (value as String?) ?? ""
    }

    @objc(__lynx_prop_config__marker_id)
    public class func __lynxPropConfigMarkerId() -> [String] {
        return ["marker-id", "setMarkerId", "NSString *"]
    }
}

File: packages/lynx-maps/android/com/sigx/maps/SigxMapMarkerUI.kt
package com.sigx.maps

import android.content.Context
import android.view.View
import com.google.android.gms.maps.model.Marker
import com.lynx.tasm.behavior.LynxContext
import com.lynx.tasm.behavior.LynxProp
import com.lynx.tasm.behavior.ui.LynxUI
import org.json.JSONObject

/**
 * Native UI for the `<sigx-map-marker>` JSX element on Android.
 *
 * A marker doesn't render its own real view — Google Maps tracks pins
 * separately via [com.google.android.gms.maps.GoogleMap.addMarker]. We
 * still need a `LynxUI<View>` because Lynx's UI tree expects every child
 * to be a UI. The native view is a zero-size hidden `View` that never
 * gets laid out because the parent `SigxMapUI` intercepts `insertChild`
 * for markers and skips the default add-to-view-hierarchy path.
 */
class SigxMapMarkerUI(context: LynxContext) : LynxUI<View>(context) {

    internal var owningMap: SigxMapUI? = null
    internal var attachedMarker: Marker? = null

    internal var coordinateLat: Double = 0.0
    internal var coordinateLng: Double = 0.0
    internal var markerTitle: String? = null
    internal var markerDescription: String? = null
    internal var markerId: String = ""

    override fun createView(context: Context): View {
        val v = View(context)
        v.visibility = View.GONE
        return v
    }

    // ── Prop setters ─────────────────────────────────────────────────────

    @LynxProp(name = "coordinate")
    fun setCoordinate(value: String?) {
        if (value.isNullOrEmpty()) return
        val obj = runCatching { JSONObject(value) }.getOrNull() ?: return
        coordinateLat = obj.optDouble("latitude", coordinateLat)
        coordinateLng = obj.optDouble("longitude", coordinateLng)
        owningMap?.markerDidUpdate(this)
    }

    @LynxProp(name = "title")
    fun setTitle(value: String?) {
        markerTitle = value
        owningMap?.markerDidUpdate(this)
    }

    @LynxProp(name = "description")
    fun setDescription(value: String?) {
        markerDescription = value
        owningMap?.markerDidUpdate(this)
    }

    @LynxProp(name = "marker-id")
    fun setMarkerId(value: String?) {
        markerId = value ?: ""
    }
}
